drivers/mspi/mspi_esp32.c

```diff
@@ -0,0 +1,690 @@
+	} else {
+		hw->ctrl.d_pol = 1;
+		hw->ctrl.q_pol = 1;
+		hw->ctrl.hold_pol = 1;
+		hw->ctrl.wp_pol = 1;
+	}
+#endif
+
+	/*
+	 * Workaround for ESP32S3 and ESP32Cx SoC's. This dummy transaction is needed
+	 * to sync CLK and software controlled CS when SPI is in mode 3
+	 */
+#if defined(CONFIG_SOC_SERIES_ESP32S3) || defined(CONFIG_SOC_SERIES_ESP32C2) ||                    \
+	defined(CONFIG_SOC_SERIES_ESP32C3) || defined(CONFIG_SOC_SERIES_ESP32C6)
+	if (config->mspi_config.num_ce_gpios && (hal_dev->mode & MSPI_CPP_MODE_3)) {
+		uint8_t src[] = {0x00};
+
+		struct mspi_xfer_packet data_packet = {
+			.dir = MSPI_TX,
+			.data_buf = (void *)src,
+			.num_bytes = sizeof(src),
+		};
+
+		struct mspi_xfer xfer = {
+			.packets = &data_packet,
+			.num_packet = 1,
+			.async = false,
+			.timeout = 100,
+			.priority = 0,
+			.hold_ce = 0,
+			.cmd_length = 0,
+			.addr_length = 0,
+		};
+
+		transfer(dev, &xfer, 0);
+	}
+#endif
+
+	return 0;
+}
+
+static int mspi_esp32_dev_config(const struct device *dev, const struct mspi_dev_id *dev_id,
+				 const enum mspi_dev_cfg_mask cfg_mask,
+				 const struct mspi_dev_cfg *mspi_dev_config)
+{
+	struct mspi_esp32_data *data = dev->data;
+	const struct mspi_esp32_config *config = dev->config;
+
+	if ((cfg_mask &
+	     (~(MSPI_DEVICE_CONFIG_IO_MODE | MSPI_DEVICE_CONFIG_CPP | MSPI_DEVICE_CONFIG_CE_NUM |
+		MSPI_DEVICE_CONFIG_CE_POL | MSPI_DEVICE_CONFIG_FREQUENCY)))) {
+		LOG_ERR("Configuration type not supported.");
+		return -ENOTSUP;
+	}
+
+	if (cfg_mask & MSPI_DEVICE_CONFIG_IO_MODE) {
+		data->mspi_dev_config.io_mode = mspi_dev_config->io_mode;
+	}
+
+	if (cfg_mask & MSPI_DEVICE_CONFIG_CPP) {
+		data->mspi_dev_config.cpp = mspi_dev_config->cpp;
+		data->hal_dev_config.mode = mspi_mode_to_esp32(mspi_dev_config->cpp);
+	}
+
+	if (cfg_mask & MSPI_DEVICE_CONFIG_CE_NUM) {
+		data->mspi_dev_config.ce_num = mspi_dev_config->ce_num;
+		data->hal_dev_config.cs_pin_id = mspi_dev_config->ce_num;
+	}
+
+	if (cfg_mask & MSPI_DEVICE_CONFIG_CE_POL) {
+		data->mspi_dev_config.ce_polarity = mspi_dev_config->ce_polarity;
+		data->hal_dev_config.positive_cs =
+			mspi_dev_config->ce_polarity == MSPI_CE_ACTIVE_HIGH;
+	}
+
+	if (cfg_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+		int ret = update_timing_config(config, data, &data->hal_dev_config.timing_conf,
+					       mspi_dev_config->freq);
+
+		if (ret != ESP_OK) {
+			LOG_ERR("Failed to calculate timing config: %d", ret);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int IRAM_ATTR mspi_esp32_transceive(const struct device *dev,
+					   const struct mspi_dev_id *dev_id,
+					   const struct mspi_xfer *xfer)
+{
+	struct mspi_esp32_data *data = dev->data;
+	const struct mspi_esp32_config *config = dev->config;
+	int ret = 0;
+
+	if (xfer->num_packet == 0 || !xfer->packets) {
+		LOG_ERR("Invalid transfer parameters");
+		return -EINVAL;
+	}
+
+	if (xfer->async) {
+		LOG_ERR("Async mode not supported");
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
```

Comment:
Is there any guarantee that no deadlock occur? wait forever for mutexes in general is a bad idea when there is no bounded latency that guarantees the mutex will be unlocked sometimes which is a source for deadlocks